src/Features/Core/Portable/FileHeaders/AbstractFileHeaderCodeFixProvider.cs

```diff
@@ -0,0 +1,237 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Formatting;
+
+namespace Microsoft.CodeAnalysis.FileHeaders
+{
+    internal abstract class AbstractFileHeaderCodeFixProvider : CodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds { get; }
+            = ImmutableArray.Create(IDEDiagnosticIds.FileHeaderMismatch);
+
+        public override FixAllProvider GetFixAllProvider()
+            => new FixAll(this);
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                context.RegisterCodeFix(
+                    new MyCodeAction(cancellationToken => GetTransformedDocumentAsync(context.Document, cancellationToken)),
+                    diagnostic);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        protected abstract AbstractFileHeaderHelper FileHeaderHelper { get; }
+
+        protected abstract SyntaxTrivia EndOfLine(string text);
+        protected abstract SyntaxTriviaList ParseLeadingTrivia(string text);
+
+        private async Task<Document> GetTransformedDocumentAsync(Document document, CancellationToken cancellationToken)
+        {
+            return document.WithSyntaxRoot(await GetTransformedSyntaxRootAsync(document, cancellationToken).ConfigureAwait(false));
+        }
+
+        private async Task<SyntaxNode> GetTransformedSyntaxRootAsync(Document document, CancellationToken cancellationToken)
+        {
+            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var root = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var options = document.Project.AnalyzerOptions.AnalyzerConfigOptionsProvider.GetOptions(tree);
+            if (!options.TryGetEditorConfigOption(CodeStyleOptions.FileHeaderTemplate, out var fileHeaderTemplate)
+                || string.IsNullOrEmpty(fileHeaderTemplate))
+            {
+                // Avoid making changes if we fail to read the expected header
+                return root;
+            }
+
+            var expectedFileHeader = fileHeaderTemplate.Replace("{fileName}", Path.GetFileName(document.FilePath));
+
+            var fileHeader = FileHeaderHelper.ParseFileHeader(root);
+            SyntaxNode newSyntaxRoot;
+            if (fileHeader.IsMissing)
+            {
+                newSyntaxRoot = AddHeader(document, root, expectedFileHeader);
+            }
+            else
+            {
+                newSyntaxRoot = ReplaceHeader(document, root, expectedFileHeader);
+            }
```

Comment:
```suggestion
            var newSyntaxRoot = fileHeader.IsMissing
                ? AddHeader(document, root, expectedFileHeader)
                :  ReplaceHeader(document, root, expectedFileHeader);
```

Reply:
I prefer the full form since the operations are distinct.